src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -162,6 +162,41 @@ static SemanticTokensHelpers()
             return (ComputeTokens(text.Lines, updatedClassifiedSpans, tokenTypesToIndex), isFinalized);
         }
 
+        private static async Task GetClassifiedSpansForDocumentAsync(
+            Document document,
+            TextSpan textSpan,
+            ArrayBuilder<ClassifiedSpan> classifiedSpans,
+            bool isRazorDoc,
+            CancellationToken cancellationToken)
+        {
+            var classificationService = document.GetRequiredLanguageService<IClassificationService>();
+
+            // Case 1 - Generated Razor documents:
+            //     In Razor, the C# syntax classifier does not run on the client. This means we need to return both
+            //     syntactic and semantic classifications.
+            // Case 2 - C# and VB documents:
+            //     In C#/VB, the syntax classifier runs on the client. This means we only need to return semantic
+            //     classifications.
+            //
+            // Ideally, Razor will eventually run the classifier on their end so we can get rid of this special
+            // casing: https://github.com/dotnet/razor-tooling/issues/5850
+            if (isRazorDoc)
+            {
+                var spans = await ClassifierHelper.GetClassifiedSpansAsync(
+                    document, textSpan, cancellationToken, removeAdditiveSpans: false, fillInClassifiedSpanGaps: false).ConfigureAwait(false);
+                classifiedSpans.AddRange(spans);
```

Comment:
so, i have a subjective question for you.   It was def me that pushed using ClassifierHelper.getClassifiedSpansAsync.  However, in order to do so you had to add a fair amount of logic to it to get what you need for the LSP case.

My overall question is which do *you* think is better of the following two options:

1. keep the PR as is, and have the extra complexity in that method.
2. just provide a totally different method that, while similar to the above, is independent and dedicated for the LSP case.

If you feel like '2' actually ends up adding/repeating too much code.  Then let's stick with what's in the PR. However, if you feel that the two methods are different enough that trying to force them to be one method (along with several booleans controlling) is just to overwrought, then maybe we should just have two.

I don't have enough context to say from reading this.  However, my hope was that you could just use hte method as is.  the need to pass all this customization logic in is making me rethink that :)

Reply:
Just tested out option 2). Unfortunately, it does feel like a substantial amount of the code is duplicated between the LSP and non-LSP methods. The only portions that differ are a) whether to remove additive spans and b) whether to fill in classified span gaps. The optional parameters don't feel great to me either, but I feel like the amount of duplication we avoid may be worth the tradeoff.

Reply:
sgtm.